mne/io/nedf/nedf.py

```diff
@@ -0,0 +1,181 @@
+    #  (500HzCh1S2, 500HzCh2S2, …, 500HzChnS2), …
+    #  (500HzCh1S5,  500HzCh2S5, …, 500HzChnS5)),
+    #  (100HzCh1S2, 100HzCh2S2, 100HzCh3S2) and so on
+
+    # dtype for the binary data block
+    dt = []
+    # dtype for a single EEG sample
+    datadt = []
+    with open(filename, 'rb') as f:
+        header = f.read(10240)
+
+    headerend = header.find(b'\0')
+    if headerend == -1:
+        raise RuntimeError('End of header null not found')
+    headerxml = ElementTree.fromstring(header[:headerend])
+    nedfversion = headerxml.findtext('NEDFversion', '')
+    if nedfversion not in ['1.3', '1.4']:
+        print('Unexpected NEDFversion, hope this works anyway')
+    if headerxml.findtext('AdditionalChannelStatus', 'OFF') != 'OFF':
+        raise RuntimeError('Unknown additional channel, aborting.')
+
+    n_acc = int(headerxml.findtext('NumberOfChannelsOfAccelerometer', 0))
+    if n_acc:
+        # expect one sample of u16 accelerometer data per block
+        dt.append(('acc', '>u2', (n_acc,)))
+
+    eegset = headerxml.find('EEGSettings')
+    if eegset is None:
+        raise RuntimeError('No EEG channels found')
+    nchantotal = int(eegset.find('TotalNumberOfChannels').text)
+    info['nchan'] = nchantotal
+    info['sfreq'] = int(eegset.find('EEGSamplingRate').text)
+    info['ch_names'] = [e.text for e in eegset.find('EEGMontage')]
+    # expect nchantotal uint24s
+    datadt.append(('eeg', 'B', (nchantotal, 3)))
+
+    info['units'] = eegset.find('EEGUnits')
+
+    if headerxml.find('STIMSettings'):
+        # 2* -> two stim samples per eeg sample
+        datadt.append(('stim', 'B', (2, nchantotal, 3)))
+
+    if 'AdditionalChannelStatus' in headerxml:
+        raise RuntimeError('Unexpected AdditionalChannelStatus')
+
+    if headerxml.findtext('stepDetails/DeviceClass', '') == 'STARSTIM':
+        print('Found Starstim, not sure how to handle this')
+
+    # Trigger data: 4 bytes in newer versions, 1 byte in older versions
+    trigger_type = '>i4' if headerxml.findtext('NEDFversion') else 'B'
+    datadt.append(('trig', trigger_type))
+    # 5 data samples per block
+    dt.append(('data', np.dtype(datadt), (5,)))
+
+    date = headerxml.findtext('StepDetails/StartDate_firstEEGTimestamp', '')
+    info['meas_date'] = datetime.datetime.utcfromtimestamp(int(date) / 1000)
+    return info, np.dtype(dt)
+
+
+def _read_nedf_eeg(filename: str):
+    """
+    Read header info and EEG data from an .nedf file
+
+    Parameters
+    ----------
+    filename : str
+        Path to the .nedf file.
+
+    Returns
+    -------
+    eeg : array, shape (n_samples, n_channels)
+        Unscaled EEG data
+    info : dict
+        Information from the file header
+    triggers : array, shape (n_annots, 2)
+        Start samples and values of each trigger
+    scale : float
+        Scaling factor for the EEG data
+    """
+    info, dt = parse_nedf_header(filename)
+
+    # to quote the original matlab implementation:
```

Comment:
Is the original implementation linked somewhere? Is it necessary to check for license compatibility, or did you write this code from scratch?

Reply:
I read through the EEGLAB plugin (https://www.neuroelectrics.com/resources/software/) to get an idea how the nedf files are structured, but other than that everything's written from scratch.

Reply:
OK, then either don't mention "the original matlab implementation" in the comment or provide a link so that people can check it out themselves.

Reply:
Done.